internal/buildkitd/bundling/main_linux.go

```diff
@@ -0,0 +1,825 @@
+// Buildkitd has been copied from https://github.com/moby/buildkit/tree/d47b46cf2a16ca80a958384282e8028285b1866d/cmd/buildkitd
+package buildkitd
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"os"
+	"os/user"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/pkg/seed"
+	"github.com/containerd/containerd/pkg/userns"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/containerd/containerd/sys"
+	sddaemon "github.com/coreos/go-systemd/v22/daemon"
+	"github.com/docker/docker/pkg/reexec"
+	"github.com/gofrs/flock"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+
+	"github.com/moby/buildkit/cache/remotecache"
+	"github.com/moby/buildkit/cache/remotecache/gha"
+	inlineremotecache "github.com/moby/buildkit/cache/remotecache/inline"
+	localremotecache "github.com/moby/buildkit/cache/remotecache/local"
+	registryremotecache "github.com/moby/buildkit/cache/remotecache/registry"
+	"github.com/moby/buildkit/worker"
+
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/cmd/buildkitd/config"
+	"github.com/moby/buildkit/control"
+	"github.com/moby/buildkit/executor/oci"
+	"github.com/moby/buildkit/frontend"
+	dockerfile "github.com/moby/buildkit/frontend/dockerfile/builder"
+	"github.com/moby/buildkit/frontend/gateway"
+	"github.com/moby/buildkit/frontend/gateway/forwarder"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/solver/bboltcachestorage"
+	"github.com/moby/buildkit/util/apicaps"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/moby/buildkit/util/appdefaults"
+	"github.com/moby/buildkit/util/archutil"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/moby/buildkit/util/grpcerrors"
+	"github.com/moby/buildkit/util/profiler"
+	"github.com/moby/buildkit/util/resolver"
+	"github.com/moby/buildkit/util/stack"
+	"github.com/moby/buildkit/util/tracing/detect"
+	_ "github.com/moby/buildkit/util/tracing/detect/jaeger"
+	_ "github.com/moby/buildkit/util/tracing/env"
+	"github.com/moby/buildkit/util/tracing/transform"
+	"github.com/moby/buildkit/version"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"go.opentelemetry.io/otel/propagation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	tracev2 "go.opentelemetry.io/otel/trace"
+	tracev1 "go.opentelemetry.io/proto/otlp/collector/trace/v1"
+
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+)
+
+func init() {
+	apicaps.ExportedProduct = "buildkit"
+	stack.SetVersionInfo(version.Version, version.Revision)
+
+	seed.WithTimeAndRand()
+	if reexec.Init() {
+		os.Exit(0)
+	}
```

Comment:
These lines are sort of scary in that they affect the entire process and rely on mysterious magic. I did a little digging:
* The actual implementation is here: https://github.com/moby/moby/blob/master/pkg/reexec/reexec.go
* It seems to be used in buildkitd to enable re-execing the buildkitd process as git plus some umask stuff? https://github.com/sipsma/buildkit/blob/61307b8f2e73d681721b51856c10cb8606e5ea70/source/git/gitsource_unix.go#L23-L23

So I think we have to leave these lines in, but can we add a comment above here saying that the line is needed for the git source implementation and link to that second code link I provided? Just to save time in the future.